### PR TITLE
perf: mpool: Implement transactionLk to ensure consistent state access in MessagePool

### PR DESCRIPTION
**Pull request by snissn**
This commit introduces a new lock transactionLk to the MessagePool struct in order to provide a consistent view of the state and protect from state changes. 

The primary objective of transactionLk is to ensure there is at most a single concurrent writer of message pool state. By holding this lock, necessary computations can be performed before making the actual state changes under the existing curTsLk. Meanwhile, other methods like GetStateNonce can still acquire read locks on curTsLk, ensuring a consistent view of the state. This approach allows writers, such as the addTs function, to rely on the transactionLk to maintain the exclusivity of a single concurrent writer during their execution.

It is worth determining if mp.transactionLk  makes mp.Lk redundant and can be removed. However we may need both as mp.Lk has a comment indicating it needs to be inside curTsLk and transactionLk has to be used outside of curTsLk

Key changes include:

- Add transactionLk to the MessagePool struct.
- Acquire and release transactionLk in functions like addTs, HeadChange, pruneExcessMessages, republishPendingMessages, and SelectMessages.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by snissn** [comment]
todo: look at failing tests

File: chain/messagepool/messagepool.go [hunk 1]
```diff
@@ -139,9 +139,14 @@ type MessagePool struct {
 
 	keyCache *lru.Cache[address.Address, address.Address]
 
-	curTsLk sync.RWMutex // DO NOT LOCK INSIDE lk
+	stateLk sync.RWMutex
 	curTs   *types.TipSet
 
+	// Use this lock if you need to transact between the TS and the Message pool in a consistent manner
+	//all changes to curTs must be wrapped in a transactionLk
+	//You do not need this lock and only need curTsLk if you want a consistent view of the state without changing the state
```
Comment (shrenujbansal) [inline chain/messagepool/messagepool.go:147]: Change name here from curTsLk to stateLk in accordance with the change on line 142

File: chain/messagepool/messagepool.go [hunk 2]
```diff
@@ -139,9 +139,14 @@ type MessagePool struct {
 
 	keyCache *lru.Cache[address.Address, address.Address]
 
-	curTsLk sync.RWMutex // DO NOT LOCK INSIDE lk
+	stateLk sync.RWMutex
```
Comment (shrenujbansal) [inline chain/messagepool/messagepool.go:142]: Could you also had a similar comment as below describing this lock?

File: chain/messagepool/messagepool.go [hunk 3]
```diff
@@ -867,7 +845,21 @@ func (mp *MessagePool) checkBalance(ctx context.Context, m *types.SignedMessage,
 	return nil
 }
 
-func (mp *MessagePool) addTs(ctx context.Context, m *types.SignedMessage, curTs *types.TipSet, local, untrusted bool) (bool, error) {
+func (mp *MessagePool) addTs(ctx context.Context, m *types.SignedMessage, local, untrusted bool) (bool, error) {
+
+	//ensures that we have a consistent view of the state
+	mp.transactionLk.Lock()
```
[the remark this answers is not in the document]
Comment (snissn) [reply chain/messagepool/messagepool.go:851]: yes!! The transactionLk is to ensure that the writer of state has a coherent view of the state, while the actual state changes happen under stateLk. Please more review, and fortunately we don't need to rush this as the network health has improved, but we definitely do want to land this for if/when the fvm apis are slow in the future.

File: chain/messagepool/messagepool.go (hunk 1 above)
Comment (snissn) [reply chain/messagepool/messagepool.go:147]: ty!!! good catch

Comment (snissn) [reply chain/messagepool/messagepool.go]: #44bbe7fdc

Comment (snissn) [reply chain/messagepool/messagepool.go]: #44bbe7fdc4271dc29218415a9a0c89d2417bab2f

Comment (snissn) [reply chain/messagepool/messagepool.go]: https://github.com/filecoin-project/lotus/commit/44bbe7fdc4271dc29218415a9a0c89d2417bab2f

File: chain/messagepool/messagepool.go (hunk 2 above)
Comment (snissn) [reply chain/messagepool/messagepool.go]: done ty!!  4d458f119

**Comment by snissn** [comment]
closing this for PR https://github.com/filecoin-project/lotus/pull/10865 which is a single squashed commit

**Pull request closed without merging**